fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -1073,6 +1064,35 @@ public void rebuildIndex(@Nonnull FDBRecordStore store) {
         });
     }
 
+    private <T> CompletableFuture<T> runWithSynchronizedRunnerAndEndSession(
+            @Nonnull SynchronizedSessionRunner newSynchronizedRunner, @Nonnull Supplier<CompletableFuture<T>> runnable) {
+        if (this.synchronizedSessionRunner == null) {
+            this.synchronizedSessionRunner = newSynchronizedRunner;
+            return AsyncUtil.composeHandle(runnable.get(), (result, ex) -> {
+                if (this.synchronizedSessionRunner == newSynchronizedRunner) {
+                    this.synchronizedSessionRunner = null;
+                } else {
+                    LOGGER.warn(KeyValueLogMessage.of("synchronizedSessionRunner was modified during the run",
+                            LogMessageKeys.SESSION_ID, newSynchronizedRunner.getSessionId(),
+                            LogMessageKeys.INDEXER_SESSION_ID, this.synchronizedSessionRunner == null ? null : this.synchronizedSessionRunner.getSessionId()));
+                }
+                return newSynchronizedRunner.endSessionAsync().thenApply(vigore -> {
+                    if (ex == null) {
+                        return result;
+                    } else {
+                        throw new RuntimeException(ex);
```

Comment:
Is there a reason this isn't `throw getRunner().getDatabase().mapAsyncToSyncException(ex);`?

I'd be curious to know if that also fixes the need to unwrap the exceptions twice in the tests.